kornia/feature/keynet.py

```diff
@@ -0,0 +1,315 @@
+import math
+from typing import Dict, List, Optional, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from kornia.filters import spatial_gradient
+from kornia.geometry.subpix import NonMaximaSuppression2d
+from kornia.geometry.transform import pyrdown
+
+from .laf import laf_from_center_scale_ori
+from .orientation import PassLAF
+
+keynet_config = {
+
+    'KeyNet_default_config':
+        {
+            # Key.Net Model
+            'num_filters': 8,
+            'num_levels': 3,
+            'kernel_size': 5,
+            # Extraction Parameters
+            'nms_size': 15,
+            'pyramid_levels': 4,
+            'up_levels': 1,
+            'scale_factor_levels': math.sqrt(2),
+            's_mult': 22,
+        },
+}
+
+urls: Dict[str, str] = {}
+urls["keynet"] = "https://github.com/axelBarroso/Key.Net-Pytorch/raw/main/model/weights/keynet_pytorch.pth"
+
+
+class feature_extractor(nn.Module):
+    """Helper class for KeyNet.
+
+    It loads both, the handcrafted and learnable blocks
+    """
+    def __init__(self):
+        super().__init__()
+
+        self.hc_block = handcrafted_block()
+        self.lb_block = learnable_block()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_hc = self.hc_block(x)
+        x_lb = self.lb_block(x_hc)
+        return x_lb
+
+
+class handcrafted_block(nn.Module):
+    """Helper class for KeyNet, it defines the handcrafted filters within the Key.Net handcrafted block."""
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+        sobel = spatial_gradient(x)
+        dx, dy = sobel[:, :, 0, :, :], sobel[:, :, 1, :, :]
+
+        sobel_dx = spatial_gradient(dx)
+        dxx, dxy = sobel_dx[:, :, 0, :, :], sobel_dx[:, :, 1, :, :]
+
+        sobel_dy = spatial_gradient(dy)
+        dyy = sobel_dy[:, :, 1, :, :]
+
+        hc_feats = torch.cat([dx,
+                              dy,
+                              dx ** 2.,
+                              dy ** 2.,
+                              dx * dy,
+                              dxy,
+                              dxy ** 2.,
+                              dxx,
+                              dyy,
+                              dxx * dyy], dim=1)
+
+        return hc_feats
+
+
+class learnable_block(nn.Module):
```

Comment:
```suggestion
class learnable_block(nn.Sequential):
```
probably more cleaner